File: nodejs-backend/src/controllers/payment.controller.js
/**
 * Payment Controller
 * Handles payment endpoints with fake/test payment simulation
 * Converted from Spring Boot PaymentController
 */

const paymentService = require('../services/payment.service');

class PaymentController {
  /**
   * GET /api/payments
   * Get all payments
   */
  async getAllPayments(req, res, next) {
    try {
      const payments = await paymentService.getAllPayments();
      res.json(payments);
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/payments/:id
   * Get payment by ID
   */
  async getPaymentById(req, res, next) {
    try {
      const payment = await paymentService.getPaymentById(req.params.id);
      if (!payment) {
        return res.status(404).json({
          success: false,
          message: 'Payment not found'
        });
      }
      res.json(payment);
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/payments/transaction/:transactionId
   * Get payment by transaction ID
   */
  async getPaymentByTransactionId(req, res, next) {
    try {
      const payment = await paymentService.getPaymentByTransactionId(req.params.transactionId);
      if (!payment) {
        return res.status(404).json({
          success: false,
          message: 'Payment not found'
        });
      }
      res.json(payment);
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/payments/user/:email
   * Get payments by user email
   */
  async getPaymentsByUser(req, res, next) {
    try {
      const payments = await paymentService.getPaymentsByUserEmail(req.params.email);
      res.json(payments);
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/payments/completed
   * Get completed payments
   */
  async getCompletedPayments(req, res, next) {
    try {
      const payments = await paymentService.getCompletedPayments();
      res.json(payments);
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/payments/stats
   * Get payment statistics
   */
  async getPaymentStats(req, res, next) {
    try {
      const stats = {
        totalRevenue: await paymentService.getTotalRevenue(),
        cardPayments: await paymentService.getPaymentCountByMethod('CARD'),
        paypalPayments: await paymentService.getPaymentCountByMethod('PAYPAL'),
        stripePayments: await paymentService.getPaymentCountByMethod('STRIPE'),
        wisePayments: await paymentService.getPaymentCountByMethod('WISE')
      };
      res.json(stats);
    } catch (error) {
      next(error);
    }
  }

  /**
   * POST /api/payments/card
   * Process credit card payment (FAKE/TEST API)
   */
  async processCardPayment(req, res, next) {
    try {
      console.log('Processing card payment request');
      const payment = await paymentService.processCardPayment(req.body);
      
      res.json({
        success: payment.status === 'COMPLETED',
        transactionId: payment.transaction_id,
        status: payment.status,
        message: payment.status === 'COMPLETED' 
          ? 'Payment processed successfully' 
          : payment.error_message,
        payment
      });
    } catch (error) {
      console.error('Card payment error:', error.message);
      res.status(400).json({
        success: false,
        error: error.message
      });
    }
  }

  /**
   * POST /api/payments/paypal/create
   * Create PayPal payment intent
   */
  async createPayPalPayment(req, res, next) {
    try {
      req.body.paymentMethod = 'PAYPAL';
      const payment = await paymentService.createExternalPayment(req.body);
      
      res.json({
        success: true,
        transactionId: payment.transaction_id,
        paymentId: payment.id,
        approvalUrl: `https://www.sandbox.paypal.com/checkoutnow?token=${payment.transaction_id}`
      });
    } catch (error) {
      console.error('PayPal create error:', error.message);
      res.status(400).json({ error: error.message });
    }
  }

  /**
   * POST /api/payments/paypal/confirm
   * Confirm PayPal payment
   */
  async confirmPayPalPayment(req, res, next) {
    try {
      const { transactionId, paypalOrderId } = req.body;
      const payment = await paymentService.confirmExternalPayment(transactionId, paypalOrderId, true);
      
      res.json({
        success: true,
        transactionId: payment.transaction_id,
        status: payment.status,
        message: 'PayPal payment confirmed successfully'
      });
    } catch (error) {
      console.error('PayPal confirm error:', error.message);
      res.status(400).json({ error: error.message });
    }
  }

  /**
   * POST /api/payments/stripe/create
   * Create Stripe payment intent
   */
  async createStripePayment(req, res, next) {
    try {
      req.body.paymentMethod = 'STRIPE';
      const payment = await paymentService.createExternalPayment(req.body);
      
      res.json({
        success: true,
        transactionId: payment.transaction_id,
        paymentId: payment.id,
        clientSecret: `pi_test_${payment.transaction_id}_secret_xxx`
      });
    } catch (error) {
      console.error('Stripe create error:', error.message);
      res.status(400).json({ error: error.message });
    }
  }

  /**
   * POST /api/payments/stripe/confirm
   * Confirm Stripe payment
   */
  async confirmStripePayment(req, res, next) {
    try {
      const { transactionId, paymentIntentId } = req.body;
      const payment = await paymentService.confirmExternalPayment(transactionId, paymentIntentId, true);
      
      res.json({
        success: true,
        transactionId: payment.transaction_id,
        status: payment.status,
        message: 'Stripe payment confirmed successfully'
      });
    } catch (error) {
      console.error('Stripe confirm error:', error.message);
      res.status(400).json({ error: error.message });
    }
  }

  /**
   * POST /api/payments/wise/create
   * Create Wise transfer
   */
  async createWiseTransfer(req, res, next) {
    try {
      req.body.paymentMethod = 'WISE';
      const payment = await paymentService.createExternalPayment(req.body);
      
      res.json({
        success: true,
        transactionId: payment.transaction_id,
        paymentId: payment.id,
        transferUrl: `https://wise.com/pay/${payment.transaction_id}`
      });
    } catch (error) {
      console.error('Wise create error:', error.message);
      res.status(400).json({ error: error.message });
    }
  }

  /**
   * POST /api/payments/wise/confirm
   * Confirm Wise transfer
   */
  async confirmWiseTransfer(req, res, next) {
    try {
      const { transactionId, transferId } = req.body;
      const payment = await paymentService.confirmExternalPayment(transactionId, transferId, true);
      
      res.json({
        success: true,
        transactionId: payment.transaction_id,
        status: payment.status,
        message: 'Wise transfer confirmed successfully'
      });
    } catch (error) {
      console.error('Wise confirm error:', error.message);
      res.status(400).json({ error: error.message });
    }
  }

  /**
   * PATCH /api/payments/:id/status
   * Update payment status
   */
  async updatePaymentStatus(req, res, next) {
    try {
      const { status } = req.body;
      const payment = await paymentService.updatePaymentStatus(req.params.id, status);
      res.json(payment);
    } catch (error) {
      res.status(400).json({ error: error.message });
    }
  }
}

module.exports = new PaymentController();

File: Back-end/src/main/java/KJX/KJX/controller/PaymentController.java
package KJX.KJX.controller;

import KJX.KJX.entity.Payment;
import KJX.KJX.service.PaymentService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/payments")
@RequiredArgsConstructor
@CrossOrigin(origins = "*")
@Slf4j
public class PaymentController {
    
    private final PaymentService paymentService;
    
    // =============== GET ENDPOINTS ===============
    
    @GetMapping
    public ResponseEntity<List<Payment>> getAllPayments() {
        return ResponseEntity.ok(paymentService.getAllPayments());
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<Payment> getPaymentById(@PathVariable Long id) {
        return paymentService.getPaymentById(id)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }
    
    @GetMapping("/transaction/{transactionId}")
    public ResponseEntity<Payment> getPaymentByTransactionId(@PathVariable String transactionId) {
        return paymentService.getPaymentByTransactionId(transactionId)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }
    
    @GetMapping("/user/{email}")
    public ResponseEntity<List<Payment>> getPaymentsByUser(@PathVariable String email) {
        return ResponseEntity.ok(paymentService.getPaymentsByUserEmail(email));
    }
    
    @GetMapping("/completed")
    public ResponseEntity<List<Payment>> getCompletedPayments() {
        return ResponseEntity.ok(paymentService.getCompletedPayments());
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> getPaymentStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalRevenue", paymentService.getTotalRevenue());
        stats.put("cardPayments", paymentService.getPaymentCountByMethod("CARD"));
        stats.put("paypalPayments", paymentService.getPaymentCountByMethod("PAYPAL"));
        stats.put("stripePayments", paymentService.getPaymentCountByMethod("STRIPE"));
        stats.put("wisePayments", paymentService.getPaymentCountByMethod("WISE"));
        return ResponseEntity.ok(stats);
    }
    
    // =============== CARD PAYMENT (FAKE API) ===============
    
    /**
     * Process credit card payment - FAKE/TEST API
     * Validates card details and simulates payment processing
     */
    @PostMapping("/card")
    public ResponseEntity<?> processCardPayment(@RequestBody Map<String, Object> paymentData) {
        try {
            log.info("Processing card payment request");
            Payment payment = paymentService.processCardPayment(paymentData);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", "COMPLETED".equals(payment.getStatus()));
            response.put("transactionId", payment.getTransactionId());
            response.put("status", payment.getStatus());
            response.put("message", "COMPLETED".equals(payment.getStatus()) 
                ? "Payment processed successfully" 
                : payment.getErrorMessage());
            response.put("payment", payment);
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Card payment error: {}", e.getMessage());
            Map<String, Object> errorResponse = new HashMap<>();
            errorResponse.put("success", false);
            errorResponse.put("error", e.getMessage());
            return ResponseEntity.badRequest().body(errorResponse);
        }
    }
    
    // =============== EXTERNAL PAYMENTS ===============
    
    /**
     * Create payment intent for PayPal
     */
    @PostMapping("/paypal/create")
    public ResponseEntity<?> createPayPalPayment(@RequestBody Map<String, Object> paymentData) {
        try {
            paymentData.put("paymentMethod", "PAYPAL");
            Payment payment = paymentService.createExternalPayment(paymentData);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("transactionId", payment.getTransactionId());
            response.put("paymentId", payment.getId());
            // In real implementation, this would return PayPal approval URL
            response.put("approvalUrl", "https://www.sandbox.paypal.com/checkoutnow?token=" + payment.getTransactionId());
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("PayPal create error: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
    
    /**
     * Confirm PayPal payment after user approval
     */
    @PostMapping("/paypal/confirm")
    public ResponseEntity<?> confirmPayPalPayment(@RequestBody Map<String, Object> confirmData) {
        try {
            String transactionId = (String) confirmData.get("transactionId");
            String paypalOrderId = (String) confirmData.get("paypalOrderId");
            
            Payment payment = paymentService.confirmExternalPayment(transactionId, paypalOrderId, true);
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "transactionId", payment.getTransactionId(),
                "status", payment.getStatus(),
                "message", "PayPal payment confirmed successfully"
            ));
        } catch (Exception e) {
            log.error("PayPal confirm error: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
    
    /**
     * Create Stripe payment intent
     */
    @PostMapping("/stripe/create")
    public ResponseEntity<?> createStripePayment(@RequestBody Map<String, Object> paymentData) {
        try {
            paymentData.put("paymentMethod", "STRIPE");
            Payment payment = paymentService.createExternalPayment(paymentData);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("transactionId", payment.getTransactionId());
            response.put("paymentId", payment.getId());
            // In real implementation, this would return Stripe client secret
            response.put("clientSecret", "pi_test_" + payment.getTransactionId() + "_secret_xxx");
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Stripe create error: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
    
    /**
     * Confirm Stripe payment
     */
    @PostMapping("/stripe/confirm")
    public ResponseEntity<?> confirmStripePayment(@RequestBody Map<String, Object> confirmData) {
        try {
            String transactionId = (String) confirmData.get("transactionId");
            String stripePaymentIntentId = (String) confirmData.get("paymentIntentId");
            
            Payment payment = paymentService.confirmExternalPayment(transactionId, stripePaymentIntentId, true);
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "transactionId", payment.getTransactionId(),
                "status", payment.getStatus(),
                "message", "Stripe payment confirmed successfully"
            ));
        } catch (Exception e) {
            log.error("Stripe confirm error: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
    
    /**
     * Create Wise transfer
     */
    @PostMapping("/wise/create")
    public ResponseEntity<?> createWiseTransfer(@RequestBody Map<String, Object> paymentData) {
        try {
            paymentData.put("paymentMethod", "WISE");
            Payment payment = paymentService.createExternalPayment(paymentData);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("transactionId", payment.getTransactionId());
            response.put("paymentId", payment.getId());
            // In real implementation, this would return Wise transfer details
            response.put("transferUrl", "https://wise.com/pay/" + payment.getTransactionId());
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            log.error("Wise create error: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
    
    /**
     * Confirm Wise transfer
     */
    @PostMapping("/wise/confirm")
    public ResponseEntity<?> confirmWiseTransfer(@RequestBody Map<String, Object> confirmData) {
        try {
            String transactionId = (String) confirmData.get("transactionId");
            String wiseTransferId = (String) confirmData.get("transferId");
            
            Payment payment = paymentService.confirmExternalPayment(transactionId, wiseTransferId, true);
            
            return ResponseEntity.ok(Map.of(
                "success", true,
                "transactionId", payment.getTransactionId(),
                "status", payment.getStatus(),
                "message", "Wise transfer confirmed successfully"
            ));
        } catch (Exception e) {
            log.error("Wise confirm error: {}", e.getMessage());
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
    
    // =============== UPDATE STATUS ===============
    
    @PatchMapping("/{id}/status")
    public ResponseEntity<?> updatePaymentStatus(@PathVariable Long id, @RequestBody Map<String, String> statusData) {
        try {
            String newStatus = statusData.get("status");
            Payment payment = paymentService.updatePaymentStatus(id, newStatus);
            return ResponseEntity.ok(payment);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }
}
